subsys/bluetooth/mesh/scheduler_srv.c

```diff
@@ -0,0 +1,749 @@
+	run_scheduler(srv);
+}
+
+static void encode_status(struct bt_mesh_scheduler_srv *srv,
+			  struct net_buf_simple *buf)
+{
+	bt_mesh_model_msg_init(buf, BT_MESH_SCHEDULER_OP_STATUS);
+	net_buf_simple_add_le16(buf, srv->status_bitmap);
+	BT_DBG("Tx: scheduler server status: %#2x", srv->status_bitmap);
+}
+
+static void encode_action_status(struct bt_mesh_scheduler_srv *srv,
+				 struct net_buf_simple *buf,
+				 uint8_t idx)
+{
+	bt_mesh_model_msg_init(buf, BT_MESH_SCHEDULER_OP_ACTION_STATUS);
+	scheduler_action_pack(buf, idx, &srv->sch_reg[idx]);
+
+	BT_DBG("Tx: scheduler server action status:");
+	BT_DBG("              index: %d", idx);
+	BT_DBG("               year: %d", srv->sch_reg[idx].year);
+	BT_DBG("              month: %#2x", srv->sch_reg[idx].month);
+	BT_DBG("                day: %d", srv->sch_reg[idx].day);
+	BT_DBG("               wday: %#2x", srv->sch_reg[idx].day_of_week);
+	BT_DBG("               hour: %d", srv->sch_reg[idx].hour);
+	BT_DBG("             minute: %d", srv->sch_reg[idx].minute);
+	BT_DBG("             second: %d", srv->sch_reg[idx].second);
+	BT_DBG("             action: %d", srv->sch_reg[idx].action);
+	BT_DBG("         transition: %d", srv->sch_reg[idx].transition_time);
+	BT_DBG("              scene: %d", srv->sch_reg[idx].scene_number);
+}
+
+static int send_scheduler_status(struct bt_mesh_model *model,
+		struct bt_mesh_msg_ctx *ctx)
+{
+	struct bt_mesh_scheduler_srv *srv = model->user_data;
+
+	BT_MESH_MODEL_BUF_DEFINE(buf, BT_MESH_SCHEDULER_OP_STATUS,
+			BT_MESH_SCHEDULER_MSG_LEN_STATUS);
+	encode_status(srv, &buf);
+
+	return model_send(model, ctx, &buf);
+}
+
+static int send_scheduler_action_status(struct bt_mesh_model *model,
+					struct bt_mesh_msg_ctx *ctx,
+					uint8_t idx)
+{
+	struct bt_mesh_scheduler_srv *srv = model->user_data;
+
+	BT_MESH_MODEL_BUF_DEFINE(buf, BT_MESH_SCHEDULER_OP_ACTION_STATUS,
+			BT_MESH_SCHEDULER_MSG_LEN_ACTION_STATUS);
+
+	encode_action_status(srv, &buf, idx);
+
+	return model_send(model, ctx, &buf);
+}
+
+static void action_set(struct bt_mesh_model *model,
+		       struct bt_mesh_msg_ctx *ctx,
+		       struct net_buf_simple *buf,
+		       bool ack)
+{
+	struct bt_mesh_scheduler_srv *srv = model->user_data;
+	uint8_t idx;
+	struct bt_mesh_schedule_entry tmp;
+
+	scheduler_action_unpack(buf, &idx, &tmp);
+
+	/* check against prohibited values */
+	if (tmp.year > BT_MESH_SCHEDULER_ANY_YEAR ||
+	    tmp.day > MAX_DAY ||
+	    tmp.hour > BT_MESH_SCHEDULER_ONCE_A_DAY ||
+	    tmp.minute > BT_MESH_SCHEDULER_ONCE_AN_HOUR ||
+	    tmp.second > BT_MESH_SCHEDULER_ONCE_A_MINUTE ||
+	    idx >= BT_MESH_SCHEDULER_ACTION_ENTRY_COUNT) {
+		return;
+	}
+
+	srv->sch_reg[idx] = tmp;
+	BT_DBG("Rx: scheduler server action index %d set, ack %d", idx, ack);
+
+	if (srv->sch_reg[idx].action < BT_MESH_SCHEDULER_SCENE_RECALL ||
+	   (srv->sch_reg[idx].action == BT_MESH_SCHEDULER_SCENE_RECALL &&
+	    srv->sch_reg[idx].scene_number != 0)) {
+		schedule_action(srv, idx);
```

Comment:
I see you check this here, but even if `action > BT_MESH_SCHEDULER_SCENE_RECALL` the `srv->sch_reg[idx]` will be overwritten. Is this on purpose?

Reply:
According to the standard client can set the entry configuration with the action `BT_MESH_SCHEDULER_NO_ACTIONS`  (0xf) and the Scheduler Server should return this entry configuration when client sends `Scheduler Action Get`.
However, there is no sense to schedule the register entry with action `no action`.

I'll add more strict action value checking

Reply:
According to the standard client can set configuration entry with `no action` and the server shall store it. The client can get this configuration back. Added more strict action checking.